### ISSUE TITLE
Ton of SERVFAIL domains

### ISSUE THREAD

**Issue by devipasigner** [TechnitiumSoftware/DnsServer#552]
Hello 👋 , I have set up technitium dns server on my home network and for the most part its excellent. However some domains are failing to be resolved and from the query log I can tell that the response codes are SERVFAIL. I am not sure why though, it seems the domains are timed out but this did not work with pihole/adguard with unbound. 

Right now they are mostly happening on apple domains, primarily the app store domains. I cannot install any apps from the apple app store on my iPhone, iPad or mac. Any idea why this would happen?

**Comment by ShreyasZare** [comment]
Thanks for the post. Check the DNS logs from the web panel to know why the DNS server is failing to resolve the domain names. Do share any error log you see so that I can help with debugging the issue.

**Comment by SoftExpert** [comment]
Hello, I think I have the same issue; here is the corresponding exception retrieved from the log. I'm using NextDNS as forward DNS service. Constantly, the autoconfig.kde.org address is not resolved. I hope the provided trace helps ...

```
[2023-02-17 12:47:51 Local] [127.0.0.1:60298] [admin] Cached zone was deleted: cdn77.org
[2023-02-17 12:47:57 Local] [127.0.0.1:60298] [admin] Cached zone was deleted: autoconfig.kde.org
[2023-02-17 12:48:33 Local] DNS Server failed to resolve the request with QNAME: autoconfig.kde.org; QTYPE: A; QCLASS: IN; Forwarders: tech-***.dns.nextdns.io:853 (45.90.28.57), tech-***.dns.nextdns.io:853 (45.90.30.57);
TechnitiumLibrary.Net.Dns.DnsClientResponseDnssecValidationException: DNSSEC validation failed due to missing DS records for owner name: cdn77.org
   at TechnitiumLibrary.Net.Dns.DnsClient.GetDSForAsync(String ownerName, DnsClass class, IReadOnlyList`1 currentDnsKeyRecords, DnsClient dnsClient, IDnsCache cache, UInt16 udpPayloadSize, DnsDatagram originalResponse, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 3013
   at TechnitiumLibrary.Net.Dns.DnsClient.FindDnsKeyForAsync(String ownerName, DnsClass class, IReadOnlyList`1 currentDnsKeyRecords, DnsClient dnsClient, IDnsCache cache, UInt16 udpPayloadSize, DnsDatagram originalResponse, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 2834
   at TechnitiumLibrary.Net.Dns.DnsClient.DnssecValidateResponseAsync(DnsDatagram response, IReadOnlyList`1 lastDSRecords, DnsClient dnsClient, IDnsCache cache, UInt16 udpPayloadSize, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 2543
   at TechnitiumLibrary.Net.Dns.DnsClient.InternalDnssecResolveAsync(DnsQuestionRecord question, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 4358
   at TechnitiumLibrary.Net.Dns.DnsClient.<>c__DisplayClass71_0.<<InternalCachedResolveQueryAsync>b__0>d.MoveNext() in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 4470
--- End of stack trace from previous location ---
   at TechnitiumLibrary.Net.Dns.DnsClient.ResolveQueryAsync(DnsQuestionRecord question, Func`2 resolveAsync) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 3908
   at TechnitiumLibrary.Net.Dns.DnsClient.InternalCachedResolveQueryAsync(DnsQuestionRecord question, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 4471
   at DnsServerCore.Dns.DnsServer.RecursiveResolveAsync(DnsQuestionRecord question, NetworkAddress eDnsClientSubnet, IReadOnlyList`1 conditionalForwarders, Boolean dnssecValidation, Boolean cachePrefetchOperation, Boolean cacheRefreshOperation, TaskCompletionSource`1 taskCompletionSource) in Z:\Technitium\Projects\DnsServer\DnsServerCore\Dns\DnsServer.cs:line 2971
```

**Comment by ShreyasZare** [comment]
> Hello, I think I have the same issue; here is the corresponding exception retrieved from the log. I'm using NextDNS as forward DNS service. Constantly, the autoconfig.kde.org address is not resolved. I hope the provided trace helps ...
> 
> ```
> [2023-02-17 12:47:51 Local] [127.0.0.1:60298] [admin] Cached zone was deleted: cdn77.org
> [2023-02-17 12:47:57 Local] [127.0.0.1:60298] [admin] Cached zone was deleted: autoconfig.kde.org
> [2023-02-17 12:48:33 Local] DNS Server failed to resolve the request with QNAME: autoconfig.kde.org; QTYPE: A; QCLASS: IN; Forwarders: tech-***.dns.nextdns.io:853 (45.90.28.57), tech-***.dns.nextdns.io:853 (45.90.30.57);
> TechnitiumLibrary.Net.Dns.DnsClientResponseDnssecValidationException: DNSSEC validation failed due to missing DS records for owner name: cdn77.org
>    at TechnitiumLibrary.Net.Dns.DnsClient.GetDSForAsync(String ownerName, DnsClass class, IReadOnlyList`1 currentDnsKeyRecords, DnsClient dnsClient, IDnsCache cache, UInt16 udpPayloadSize, DnsDatagram originalResponse, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 3013
>    at TechnitiumLibrary.Net.Dns.DnsClient.FindDnsKeyForAsync(String ownerName, DnsClass class, IReadOnlyList`1 currentDnsKeyRecords, DnsClient dnsClient, IDnsCache cache, UInt16 udpPayloadSize, DnsDatagram originalResponse, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 2834
>    at TechnitiumLibrary.Net.Dns.DnsClient.DnssecValidateResponseAsync(DnsDatagram response, IReadOnlyList`1 lastDSRecords, DnsClient dnsClient, IDnsCache cache, UInt16 udpPayloadSize, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 2543
>    at TechnitiumLibrary.Net.Dns.DnsClient.InternalDnssecResolveAsync(DnsQuestionRecord question, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 4358
>    at TechnitiumLibrary.Net.Dns.DnsClient.<>c__DisplayClass71_0.<<InternalCachedResolveQueryAsync>b__0>d.MoveNext() in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 4470
> --- End of stack trace from previous location ---
>    at TechnitiumLibrary.Net.Dns.DnsClient.ResolveQueryAsync(DnsQuestionRecord question, Func`2 resolveAsync) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 3908
>    at TechnitiumLibrary.Net.Dns.DnsClient.InternalCachedResolveQueryAsync(DnsQuestionRecord question, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 4471
>    at DnsServerCore.Dns.DnsServer.RecursiveResolveAsync(DnsQuestionRecord question, NetworkAddress eDnsClientSubnet, IReadOnlyList`1 conditionalForwarders, Boolean dnssecValidation, Boolean cachePrefetchOperation, Boolean cacheRefreshOperation, TaskCompletionSource`1 taskCompletionSource) in Z:\Technitium\Projects\DnsServer\DnsServerCore\Dns\DnsServer.cs:line 2971
> ```

Thanks for the error log. This error just means that your upstream server is not responding correctly for DNSSEC requests and thus the domain is failing to resolve due to DNSSEC validation failure.

**Comment by devipasigner** [comment]
Weirdly enough this issue was caused by the advanced blocking app. As soon as I removed it and just put the same blocklists in the main blocking section in the UI, the blocking rate went way higher and servfail disappeared

**Comment by ShreyasZare** [comment]
> Weirdly enough this issue was caused by the advanced blocking app. As soon as I removed it and just put the same blocklists in the main blocking section in the UI, the blocking rate went way higher and servfail disappeared

It could be just a coincidence since the advanced blocking app does not perform any domain resolution. It only blocks domain names that are configured and does not affect DNSSEC validation on the DNS server itself.

The only way to find out the reason for domain names failing to resolve is to check the DNS logs which will log error with reason why the domain failed to resolve.

**Comment by SoftExpert** [comment]
> > Hello, I think I have the same issue; here is the corresponding exception retrieved from the log. I'm using NextDNS as forward DNS service. Constantly, the autoconfig.kde.org address is not resolved. I hope the provided trace helps ...
> > ```
> > [2023-02-17 12:47:51 Local] [127.0.0.1:60298] [admin] Cached zone was deleted: cdn77.org
> > [2023-02-17 12:47:57 Local] [127.0.0.1:60298] [admin] Cached zone was deleted: autoconfig.kde.org
> > [2023-02-17 12:48:33 Local] DNS Server failed to resolve the request with QNAME: autoconfig.kde.org; QTYPE: A; QCLASS: IN; Forwarders: tech-***.dns.nextdns.io:853 (45.90.28.57), tech-***.dns.nextdns.io:853 (45.90.30.57);
> > TechnitiumLibrary.Net.Dns.DnsClientResponseDnssecValidationException: DNSSEC validation failed due to missing DS records for owner name: cdn77.org
> >    at TechnitiumLibrary.Net.Dns.DnsClient.GetDSForAsync(String ownerName, DnsClass class, IReadOnlyList`1 currentDnsKeyRecords, DnsClient dnsClient, IDnsCache cache, UInt16 udpPayloadSize, DnsDatagram originalResponse, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 3013
> >    at TechnitiumLibrary.Net.Dns.DnsClient.FindDnsKeyForAsync(String ownerName, DnsClass class, IReadOnlyList`1 currentDnsKeyRecords, DnsClient dnsClient, IDnsCache cache, UInt16 udpPayloadSize, DnsDatagram originalResponse, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 2834
> >    at TechnitiumLibrary.Net.Dns.DnsClient.DnssecValidateResponseAsync(DnsDatagram response, IReadOnlyList`1 lastDSRecords, DnsClient dnsClient, IDnsCache cache, UInt16 udpPayloadSize, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 2543
> >    at TechnitiumLibrary.Net.Dns.DnsClient.InternalDnssecResolveAsync(DnsQuestionRecord question, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 4358
> >    at TechnitiumLibrary.Net.Dns.DnsClient.<>c__DisplayClass71_0.<<InternalCachedResolveQueryAsync>b__0>d.MoveNext() in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 4470
> > --- End of stack trace from previous location ---
> >    at TechnitiumLibrary.Net.Dns.DnsClient.ResolveQueryAsync(DnsQuestionRecord question, Func`2 resolveAsync) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 3908
> >    at TechnitiumLibrary.Net.Dns.DnsClient.InternalCachedResolveQueryAsync(DnsQuestionRecord question, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 4471
> >    at DnsServerCore.Dns.DnsServer.RecursiveResolveAsync(DnsQuestionRecord question, NetworkAddress eDnsClientSubnet, IReadOnlyList`1 conditionalForwarders, Boolean dnssecValidation, Boolean cachePrefetchOperation, Boolean cacheRefreshOperation, TaskCompletionSource`1 taskCompletionSource) in Z:\Technitium\Projects\DnsServer\DnsServerCore\Dns\DnsServer.cs:line 2971
> > ```
> 
> Thanks for the error log. This error just means that your upstream server is not responding correctly for DNSSEC requests and thus the domain is failing to resolve due to DNSSEC validation failure.

Humm, not sure if this is relevant, but when trying to isolate the case that produced the error log, I saw several times that, in addition to the logged exceptions, I also have a cdn77.org domain appearing, which also fails to be solved.
Since NextDNS is quite a large provider, I'm not sure that their response is incorrect - I would rather suspect an incompatible response, which means that perhaps there are some improvements to be done in the parsing of the response ?

**Comment by ShreyasZare** [comment]
> > > Hello, I think I have the same issue; here is the corresponding exception retrieved from the log. I'm using NextDNS as forward DNS service. Constantly, the autoconfig.kde.org address is not resolved. I hope the provided trace helps ...
> > > ```
> > > [2023-02-17 12:47:51 Local] [127.0.0.1:60298] [admin] Cached zone was deleted: cdn77.org
> > > [2023-02-17 12:47:57 Local] [127.0.0.1:60298] [admin] Cached zone was deleted: autoconfig.kde.org
> > > [2023-02-17 12:48:33 Local] DNS Server failed to resolve the request with QNAME: autoconfig.kde.org; QTYPE: A; QCLASS: IN; Forwarders: tech-***.dns.nextdns.io:853 (45.90.28.57), tech-***.dns.nextdns.io:853 (45.90.30.57);
> > > TechnitiumLibrary.Net.Dns.DnsClientResponseDnssecValidationException: DNSSEC validation failed due to missing DS records for owner name: cdn77.org
> > >    at TechnitiumLibrary.Net.Dns.DnsClient.GetDSForAsync(String ownerName, DnsClass class, IReadOnlyList`1 currentDnsKeyRecords, DnsClient dnsClient, IDnsCache cache, UInt16 udpPayloadSize, DnsDatagram originalResponse, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 3013
> > >    at TechnitiumLibrary.Net.Dns.DnsClient.FindDnsKeyForAsync(String ownerName, DnsClass class, IReadOnlyList`1 currentDnsKeyRecords, DnsClient dnsClient, IDnsCache cache, UInt16 udpPayloadSize, DnsDatagram originalResponse, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 2834
> > >    at TechnitiumLibrary.Net.Dns.DnsClient.DnssecValidateResponseAsync(DnsDatagram response, IReadOnlyList`1 lastDSRecords, DnsClient dnsClient, IDnsCache cache, UInt16 udpPayloadSize, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 2543
> > >    at TechnitiumLibrary.Net.Dns.DnsClient.InternalDnssecResolveAsync(DnsQuestionRecord question, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 4358
> > >    at TechnitiumLibrary.Net.Dns.DnsClient.<>c__DisplayClass71_0.<<InternalCachedResolveQueryAsync>b__0>d.MoveNext() in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 4470
> > > --- End of stack trace from previous location ---
> > >    at TechnitiumLibrary.Net.Dns.DnsClient.ResolveQueryAsync(DnsQuestionRecord question, Func`2 resolveAsync) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 3908
> > >    at TechnitiumLibrary.Net.Dns.DnsClient.InternalCachedResolveQueryAsync(DnsQuestionRecord question, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 4471
> > >    at DnsServerCore.Dns.DnsServer.RecursiveResolveAsync(DnsQuestionRecord question, NetworkAddress eDnsClientSubnet, IReadOnlyList`1 conditionalForwarders, Boolean dnssecValidation, Boolean cachePrefetchOperation, Boolean cacheRefreshOperation, TaskCompletionSource`1 taskCompletionSource) in Z:\Technitium\Projects\DnsServer\DnsServerCore\Dns\DnsServer.cs:line 2971
> > > ```
> > 
> > 
> > Thanks for the error log. This error just means that your upstream server is not responding correctly for DNSSEC requests and thus the domain is failing to resolve due to DNSSEC validation failure.
> 
> Humm, not sure if this is relevant, but when trying to isolate the case that produced the error log, I saw several times that, in addition to the logged exceptions, I also have a cdn77.org domain appearing, which also fails to be solved. Since NextDNS is quite a large provider, I'm not sure that their response is incorrect - I would rather suspect an incompatible response, which means that perhaps there are some improvements to be done in the parsing of the response ?

ServerFailure is a generic error response and it could mean anything from your Internet not working to a bug in the code. The only way to find what went wrong is by looking at the logs. Since you have a lot of ServerFailure responses being generated then there should be a lot of error logs too generated for them. 

Till now there is only one error log posted here which is just telling that the upstream server didn't return DS record or proof of its absence. This can happen sometime even when you use upstream servers like Cloudflare and usually a retry after a while resolves the issue. 

Right now, you can test the domain "autoconfig.kde.org" using the DNS Client tab on the web panel with DNSSEC validation enabled and selecting various upstream servers like Cloudflare, Google, Quad9, etc. You can also enter NextDNS endpoint that you have in there and test if its resolving well with DNSSEC validation. If you see any errors in there then do post them here. The DNS Client will show extended DNS errors for common issues which gives clues on what went wrong. Do post any error response you see with the tests. Without any error logs or DNS client response output there is nothing I can do.

**Comment by SoftExpert** [comment]
OK, I will test as you requested (at the end of the day); I just upgraded to v 11.0, so I will perform the tests with this version, if that's OK for you ...

**Comment by SoftExpert** [comment]
Sorry for the delay; I had troubles updating yesterday dotnet packages to the latest version and it took a little more than expected to sort it out.

I performed a battery of tests this morning.
Turns out that _autoconfig.kde.org_ (and others like it) is in fact an alias towards _1329036395.rsc.cdn77.org_.
NextDNS allows to activate some filters; amongst the filter lists I enabled, there is ListeFR; one of the filters in that list `||cdn77.org^$popup,third-party` is a bit too generic and, thus, the aliasing mechanism did not work, since the whole _cdn77.org_ domain is blacklisted. As soon as I added _cdn77.org_ to the exceptions whitelist, the aliasing mechanism worked correctly and _autoconfig.kde.org_ is resolved as expected.
In conclusion, **DnsServer** works properly - the issue I had is related to excessive filtering upstream. Sorry for the false alert, but I think we all can learn something from it.

**Comment by ShreyasZare** [comment]
> Sorry for the delay; I had troubles updating yesterday dotnet packages to the latest version and it took a little more than expected to sort it out.
> 
> I performed a battery of tests this morning. Turns out that _autoconfig.kde.org_ (and others like it) is in fact an alias towards _1329036395.rsc.cdn77.org_. NextDNS allows to activate some filters; amongst the filter lists I enabled, there is ListeFR; one of the filters in that list `||cdn77.org^$popup,third-party` is a bit too generic and, thus, the aliasing mechanism did not work, since the whole _cdn77.org_ domain is blacklisted. As soon as I added _cdn77.org_ to the exceptions whitelist, the aliasing mechanism worked correctly and _autoconfig.kde.org_ is resolved as expected. In conclusion, **DnsServer** works properly - the issue I had is related to excessive filtering upstream. Sorry for the false alert, but I think we all can learn something from it.

Thanks for the response. This should help anyone having similar issue.

If the upstream is set to block domain names then the local DNS server's DNSSEC validation is going to fail to validate the response.
